### PR TITLE
Add primary key to selectors table

### LINKED ISSUE #10585: Add primary key to selectors table
Percona-XtraDB-Cluster prohibit insert on table without primary key. And if adding primary key also help to select specific row.
selectors_tbl_error

### PR DESCRIPTION
**Pull request by zhangbutao**
Percona-XtraDB-Cluster prohibit insert on table without primary key. And if adding primary key also help to select specific row.
Fixes #10585

## Documentation

(x) No documentation is needed.

## Release notes

(x) No release notes entries required.

### CODE REVIEW

File: plugin/trino-resource-group-managers/src/main/java/io/trino/plugin/resourcegroups/db/ResourceGroupsDao.java [hunk 1]
```diff
@@ -69,6 +69,7 @@
     List<SelectorRecord> getSelectors(@Bind("environment") String environment);
 
     @SqlUpdate("CREATE TABLE IF NOT EXISTS selectors (\n" +
+            "  id BIGINT NOT NULL AUTO_INCREMENT PRIMARY KEY,\n" +
```
Comment (kokosing) [inline plugin/trino-resource-group-managers/src/main/java/io/trino/plugin/resourcegroups/db/ResourceGroupsDao.java:72]: I think we should use https://flywaydb.org/ to handle the schema changes. Also this is very vendor specific syntax. I am not sure what RDBMS vendors we support here.

Comment (hashhar) [reply plugin/trino-resource-group-managers/src/main/java/io/trino/plugin/resourcegroups/db/ResourceGroupsDao.java:72]: See in-flight PR https://github.com/trinodb/trino/pull/9812 for Flyway.

I beleive we plan to support Postgres, MySQL and Oracle only.

Comment (butaozhang) [reply plugin/trino-resource-group-managers/src/main/java/io/trino/plugin/resourcegroups/db/ResourceGroupsDao.java]: I have changed the syntax of primary key to make more generic. I don't know if it's a reasonable fix here, but i see resource_groups table also has similar primary key syntax.

File: plugin/trino-resource-group-managers/src/main/resources/db/migration/mysql/V3__add_selectors.sql [hunk 2]
```diff
@@ -1,10 +1,12 @@
 CREATE TABLE IF NOT EXISTS selectors (
+     id BIGINT NOT NULL AUTO_INCREMENT,
```
Comment (findepi) [inline plugin/trino-resource-group-managers/src/main/resources/db/migration/mysql/V3__add_selectors.sql:2]: I believe these files should never be mutated. @hashhar can we have a test?
For example use previous version to populate the DB state and use current to update it?

Comment (butaozhang) [reply plugin/trino-resource-group-managers/src/main/resources/db/migration/mysql/V3__add_selectors.sql:2]: Firstly,  these tables will not be changed after trino first startup unless updated  artificially. And if we update table by adding primary key, mysql will automatically update the historical rows with adding the primary key value. So , there is nothing effect on  previous version DB state.

Comment (findepi) [reply plugin/trino-resource-group-managers/src/main/resources/db/migration/mysql/V3__add_selectors.sql:2]: The whole https://github.com/trinodb/trino/tree/master/plugin/trino-resource-group-managers/src/main/resources/db/migration/mysql is used with Flyway, it's a reply-log of DB schema changes, and so individual changes shouldn't change.

@hashhar does Flyway record checksum of the changes files, would it detect an accidental modification?

Comment (kokosing) [reply plugin/trino-resource-group-managers/src/main/resources/db/migration/mysql/V3__add_selectors.sql:2]: What happens then?

Comment (findepi) [reply plugin/trino-resource-group-managers/src/main/resources/db/migration/mysql/V3__add_selectors.sql:2]: > > does Flyway record checksum of the changes files, would it detect an accidental modification?
> 
> It does.

@aalbu  we need a test ensuring no accidental modifications are done.
Could be as benign as someone reformatting large number of files, or Flyway changes checksum result in a corner case. https://github.com/trinodb/trino/pull/10586#discussion_r784099903 would catch any such issue, let's plan on doing it.

cc @hashhar

Comment (butaozhang) [reply plugin/trino-resource-group-managers/src/main/resources/db/migration/mysql/V3__add_selectors.sql:2]: Flyway will verify checksum of the  table beginning with V. If I change the V3__add_selectors.sql file and upgrade trino to this changed version,  trino server start failed  with a checksum exception.  
But Flyway will not verify checksum of the  table beginning with R. If i rename V3__add_selectors.sql to R__add_selectors.sql, and upgrade trino to this changed version , it just print a warn log '**_WARN    main    org.flywaydb.core.internal.sqlscript.DefaultSqlScriptExecutor   DB: Table 'selectors' already exists (SQL State: 42S01 - Error Code: 1050)_**' in server.log. And the trino server start successfully.
So, I wonder if we could rename the V Flyway table to R table?

**Comment by zhangbutao** [comment]
If adding primary key is a reasonable optimization, how  to do it? Should i add a new sql file named like **V6__add_primary_key_selectors.sql** with contents _alter table  selectors add id int unsigned not Null auto_increment primary key_ ? cc @findepi @hashhar

**Comment by hashhar** [comment]
@zhangbutao Rebase on current master to include changes merged in https://github.com/trinodb/trino/pull/9812.

Add all migrations as new files (and maybe for all supported databases - wdyt @nineinchnick @aalbu ).

Separate we'll look into adding tests which ensure that changes to older migrations are caught.

**Comment by findepi** [comment]
> Add all migrations as new files (and maybe for all supported databases - wdyt @nineinchnick @aalbu ),

👍 

> now only Flyway is used for managing database state so you don't need the changes to the Jdbi objects.

do we still need those jdbi annotations?

> Separate we'll look into adding tests which ensure that changes to older migrations are caught.


@nineinchnick @aalbu @hashhar do we have an issue?

**Comment by aalbu** [comment]
> > Separate we'll look into adding tests which ensure that changes to older migrations are caught.
> 
> @nineinchnick @aalbu @hashhar do we have an issue?

https://github.com/trinodb/trino/issues/10776

**Comment by aalbu** [comment]
> Add all migrations as new files (and maybe for all supported databases - wdyt @nineinchnick @aalbu )

Agreed, we should mirror the changes for all supported databases.

**Comment by zhangbutao** [comment]
I have added new migrations files for all supported databases and no changes to the Jdbi objects. What else need I do?   I have sent cla email to **cla@trino.io**  , please check. thx
 cc @hashhar @findepi @aalbu

**Review by nineinchnick — APPROVED** [review]

File: plugin/trino-resource-group-managers/src/main/resources/db/migration/postgresql/V5__add_primary_key_to_selectors.sql [hunk 3]
```diff
@@ -0,0 +1 @@
+ALTER TABLE selectors ADD COLUMN id BIGSERIAL PRIMARY KEY;
```
Comment (hashhar) [inline plugin/trino-resource-group-managers/src/main/resources/db/migration/postgresql/V5__add_primary_key_to_selectors.sql:1]: You can also use GENERATED since PostgreSQL < 10 isn't supported anymore and serials have some known issues.

```suggestion
ALTER TABLE selectors ADD COLUMN id BIGINT PRIMARY KEY GENERATED ALWAYS AS IDENTITY;
```

**Review by hashhar — APPROVED** [review]
nit: add newlines to all migrations.

LGTM otherwise.

**Comment by hashhar** [comment]
@cla-bot check

File: plugin/trino-resource-group-managers/src/main/resources/db/migration/postgresql/V5__add_primary_key_to_selectors.sql (hunk 3 above)
Comment (zhangbutao) [reply plugin/trino-resource-group-managers/src/main/resources/db/migration/postgresql/V5__add_primary_key_to_selectors.sql]: @hashhar Thanks for your suggest! Done!

**Comment by hashhar** [comment]
@cla-bot check

**Comment by zhangbutao** [comment]
Gentle ping. Can someone merge this request?

File: plugin/trino-resource-group-managers/src/main/resources/db/migration/postgresql/V5__add_primary_key_to_selectors.sql (hunk 3 above)
Comment (zhangbutao) [reply plugin/trino-resource-group-managers/src/main/resources/db/migration/postgresql/V5__add_primary_key_to_selectors.sql]: > You can also use GENERATED since PostgreSQL < 10 isn't supported anymore and serials have some known issues.

> 
But unit test will faile when using GENERATED:
`
Caused by: org.postgresql.util.PSQLException: ERROR: syntax error at or near "GENERATED"
  Position: 56
`

Comment (hashhar) [reply plugin/trino-resource-group-managers/src/main/resources/db/migration/postgresql/V5__add_primary_key_to_selectors.sql]: You need to change the version being used for tests at https://github.com/trinodb/trino/blob/d72b25ece9c721291d0d4d5b0f32be82400d8990/plugin/trino-resource-group-managers/src/test/java/io/trino/plugin/resourcegroups/db/TestDbResourceGroupsPostgresqlFlywayMigration.java#L27

9.x isn't supported anymore and we've already moved other tests to use 10.

Comment (zhangbutao) [reply plugin/trino-resource-group-managers/src/main/resources/db/migration/postgresql/V5__add_primary_key_to_selectors.sql]: I see. thx. I use pg 10.0 to test.

**Review by hashhar — APPROVED** [review]
